### ISSUE TITLE
[EX-40] review script

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#36]
_David Patrick on 2009-07-15T14:31:45Z says:_ 

 Straight-on task-doing, starting with the most important/urget things first, is what taskwarrior is about, along with the massive stress reduction that comes from knowing you're not forgetting anything, HOWEVER, if all of your focus is at the top of your lists, then it's still possible for some tasks (those with no date, or project, or tags etc.) to slip through the cracks. 
The review script will take the user through a comprehensive set of house-cleaning actions, from looking for stragglers, to identifying problem tasks, to repair of broken tasks.

see: [[review]]
     "Forums » development & feature ideas » GTD Review":http://taskwarrior.org/boards/6/topics/show/1085
"Forums » development & feature ideas » Implement task review":http://taskwarrior.org/boards/6/topics/2021
"task_review-1.0, First release of task-review script by Nikola Petrov":http://code.google.com/p/taskwarrior-utils/downloads/list (needs updating for 2.0 syntax)

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2009-07-15T14:31:45Z
Modified: 2017-01-17T02:50:57Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-04-18T05:08:55Z says:_ 

 Should be implemented as an external script.

**Comment by taskwarrior** [comment]
_David Patrick on 2011-06-18T02:30:53Z says:_ 

 follows 2.0 enhancements

**Comment by taskwarrior** [comment]
_David Patrick on 2012-03-17T15:14:05Z says:_ 

 After IRC discussions, it was decided that the most direct way to achieve this is to start with an external wrapper-script that executes "review elements" (single items) sequentially from a standardised directory, for which would suggest ~/.task/review/

The wrapper, task-review, without options should execute each script in ../review/ in alpha-numeric order, and to that end I would propose a review element naming convention something like

 ~/.task/review/04_noproject-0.7

leading numeric drives default review order, but the name part (between "_" and "-") can be used to execute just that script with 

 task-review noproject

The wrapper part (I think) needs to handle the interactivity, whereas the script can handle the message and the related list, we will want a fast and easy way to fix the problems. Using the example above as an imaginary case, the "noproject" script might say

 These 17 tasks have no project assigned to them, would you like to assign one now?
(y)es, (n)o, (q)uit

(q)uit interrupts the review script 
(n)o moves on the the next item, and 
(y)es could operate in one of two ways; list-mode or one-at-a-time

 (list-mode) 
which tasks? (enter IDs): [73,221,217,72]
what project? : [bazbat]
4 tasks assigned to "project:bazbat" 
These 13 tasks have no project assigned to them, would you like to assign one now?
(y)es, (n)o, (q)uit
   (repeating till there are no noproject tasks left, and then on to the next item)

 (one-at-a-time-mode) tasks are displayed via info command, with a wider range of one-key options like
(d)one, (D)elete, (n)ext, (p)roject, (q)uit

stepping through the noproject list one-at-a-time

**Comment by taskwarrior** [comment]
_Cory Donnelly on 2012-03-18T13:02:36Z says:_ 

 I think you're right, this would make for a great extension.

**Comment by taskwarrior** [comment]
_David Patrick on 2012-08-13T20:17:49Z says:_ 

 Review redux.
We know how important the review process is to the modern task warrior, and if we look at the essentials, and the working parts we already have, I think we can cook something up.

Let's make a review script that can act as a model for following scripts, the essential *weekly* is a good target, and let's do one that iterates by project, extending quick options and helpful prompts. 

The core question is; what constitutes a thorough and effective weekly review?".

this? http://zenhabits.net/weekly-review-key-to-gtd-and-achieving/

yeah, I think so! and I'd like to use this page as reference for an eventual review.weekly.conf, 
and figure out the interactivity, 
and all that, 
but what do _you_ think?

**Comment by taskwarrior** [comment]
_David Patrick on 2012-08-14T20:10:47Z says:_ 

 * The way I see it, a series of configurable sub-routines are managed in a review.<name>.conf file,- and it is run by the command *task [filter] review [name]*. 
[snip]
NOTE: most of this needs re-working due to new thinking on the subject

well and good, but I'm still confused about what language the script part of the script is written in. It could be Lua, python, bash, C++, or freakin DOS batch files, I donno. Personally, anything more than 20 lines of BASIC is over my head ;-) but Nikola had a go in python (http://taskwarrior.org/boards/6/topics/2021#message-2050) and it would seem a popular choice..

The other thing that occurred to me is that because of the level of interactivity, and the frequent use of refined tw commands, that it makes sense that the review scripts should generally be run in the twshell (coined) rather than the wilds of whatever xterm you use. The trouble with that is that the shell is not really in tip-top shape, that is to say barely usable. 

so maybe these two things go hand-in-hand, and the twshell makes it easy to sequence tweaked reports and prompts. One real benefit is that the deviation from command-line usage is lower from an enhanced shell, than is it from and interactive "one-key-response" shell script. By just leveraging the key-stroke-savings possible in twshell, existing cli syntax is respected, and no a)dd, d)one, D)elete, deleg)ate nonsense need take place. That's right, no big key-binding alternate modality, just cli syntax enhanced by the shell. This re-enforces cli-use and eliminates the headache of inventing an interactive mode. The current command-line usage is actually concise enough not to need a set of short-cuts, just shell navigation.

Last idea (for the afternoon) on the subject of review script, while the central review script might be *weekly*, we might start with review.features.conf, that walks through the many taskwarrior features, from logo + version, through the different reports, and color options, and a glimpse of the documentation, all by running tw commands, on live data and echoing text in the (new and improved) shell. Just run

 % tw review features

for a guided tour of taskwarrior features

**Comment by taskwarrior** [comment]
_David Patrick on 2013-10-02T21:00:30Z says:_ 

 UPDATE; The review process is now handled nicely by *trev*
https://github.com/nocejo/trev
and future plans include refinement of trev and improved integration with tw.

Make sure you have perl-readline (as per docs) and run trev.pl to see what I mean!

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2017-01-17T02:50:57Z says:_ 

 Implemented in tasksh 1.1.0.
